Step2 - 지하철 노선 기능

ssamzag opened the pull request:
안녕하세요!!  2단계 미션 지하철 노선 기능 리뷰 요청드립니다.
막 jpa, 스프링을 공부해가는 과정이라 기능구현보다 @sql 어너테이션 사용에 시간을 더 쏟은 것 같은 느낌이네요.. 😢  (경로문제, h2 insert 쿼리에 " 대신 ' 를 써야 한다거나..)

몇 가지 질문이 있습니다.
1. 최초 개발 완료 후 커밋할 때마다 어느정도의 단위로 커밋을 해야할 지 항상 고민이 됩니다. 이번 미션 같은 경우 크게 노선기능 / 관련파일 / 인수테스트로 나눴는데요. 어느 정도 단위로 커밋을 하면 적당한지 고민입니다.
2. 커밋 메시지에 feat, refactor, 등 프리픽스를 붙이는데 지금처럼 파일 이름도 함께 붙이는 게 좋은걸까요?  파일이 많은 경우 여러줄로 쓰기는 합니다.
3. 이번 과정에서 파라미터 인수로 3개이상은 쓰면 안 된다고 했는데, 테스트 코드 안에서 사용하는 메서드는 편의상 그 이상을 사용하고 있습니다. HashMap 객체만 넘기려니 코드가 아래로 길어지고, class로 빼려니 생성자 인자가 또 늘어나고.. 테스트에서도 이 원칙을 지켜야 하는걸까요?


이번 리뷰도 잘 부탁드립니다!! 항상 감사합니다.

ssamzag committed:
feat: 클래스 파일추가

ssamzag committed:
docs(README.md): 구현목록 업데이트

ssamzag committed:
chore: sql 파일 추가

ssamzag committed:
feat: 지하철 노선 RestApi 추가(추가/삭제/조회/수정 등)

ssamzag committed:
test(LineAcceptanceTest.java): 지하철 노선 인수 테스트 추가

ssamzag committed:
test(StationAcceptanceTest.java): sql 초기화 어노테이션 추가(truncate)

seokhongkim commented on src/main/java/nextstep/subway/domain/Line.java:
```diff
@@ -0,0 +1,73 @@
+    @ManyToOne(fetch = FetchType.LAZY)
+    private Station downStation;
+
+    private int distance;
```
`distance` 필드만 컬럼 설정을 하시지 않은 이유가 있을까요?

ssamzag replied:
거리값이 없는 라인은 있으면 안 되겠네요.. null 제약사항을 넣을게요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public LineResponse findById(Long id) {
+        Optional<Line> lines = lineRepository.findById(id);
```
단건인데 `lines` 라는 이름이 적절하지 않은 것 같아요.

ssamzag replied:
line으로 바꾸도록 할게요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+        return lines.map(LineResponse::of).orElse(null);
+    }
+
+    public Line changeLineById(Long id, LineRequest lineRequest) {
```
`name`, `color` 만 사용하는데 `LineRequest` 를 이용하네요.
다른 개발자가 수정 요청 API는 `LineRequest` 의 나머지 필드에 값이 들어오지 않는다는 것을 쉽게 알 수 있는 방법이 있을까요?

ssamzag replied:
단순하게 parameter를 두 개 던지는 게 좋겠네요..

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Transactional
+    public LineResponse saveLine(LineRequest lineRequest) {
+        Station upStation = stationRepository.getById(lineRequest.getUpStationId());
```
존재하지 않는 `station` 을 이용해서 `line` 을 만드는 경우 어떻게 되는지 테스트 해보셨나요?
이 코드에서 존해하지 않는 `station` 을 이용할 때 어떻게 동작할 것으로 기대하셨나요?
`getById` 는 유의해서 사용할 필요가 있는데, 존재하지 않는 데이터의 키로 엔티티를 가져올 때 어떻게 동작하는지 테스트 해보셨는지 궁금합니다.

ssamzag replied:
여러 예외 사항을 염두해두고 다시 작성하도록 할게요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+    public Line changeLineById(Long id, LineRequest lineRequest) {
+        Optional<Line> line = lineRepository.findById(id);
+
+        if (!line.isPresent()) {
```
`null` 을 반환하는 것이 업데이트에 실패했다는 의미로도 사용되는 것 같아요.
예외를 이용해서 없는 데이터임을 전달할 수 있는 방법을 사용해보는 것은 어떨까요?

ssamzag replied:
스프링에 예외를 이용한 방법이 있었군요!! @ExceptionHandler를 활용하도록 해보겠습니다!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Transactional
+    public void deleteLineById(Long id) {
+        lineRepository.deleteById(id);
```
존재하지 않는 데이터를 삭제하려고 하는 경우 어떤 일이 발생할까요?

ssamzag replied:
500에러가 떨어질 것 같습니다. 이럴 땐 명시적으로 400 예외를 발생시키는 게 좋을 것 같긴하네요..

seokhongkim commented on src/main/java/nextstep/subway/domain/LineRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface LineRepository extends JpaRepository<Line, Long> {
+    @Override
+    List<Line> findAll();
```
`findAll` 을 오버라이드하신 이유가 있을까요?

ssamzag replied:
필요가 없네요. 삭제하도록 할게요!

seokhongkim commented on src/main/java/nextstep/subway/domain/LineRepository.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+import java.util.Optional;
+
+public interface LineRepository extends JpaRepository<Line, Long> {
```
서비스는 명시적으로 `@Service` 어노테이션을 사용하셨는데, 레포지토리에는 `@Repository` 어노테이션을 사용하시지 않은 이유가 궁금합니다.

ssamzag replied:
깜빡했네요. 넣도록 할게요!!

seokhongkim commented on src/main/java/nextstep/subway/dto/LineRequest.java:
```diff
@@ -0,0 +1,37 @@
+        return distance;
+    }
+
+    public Line to(Station upStation, Station downStation) {
```
`Line` 으로 변경하기 위해 `Station` 을 다시 전달해줘야 하는데, 전달받는 `Station` 과 Dto 에 있는 `upStationId` 가 동일 데이터라는 보장은 어떻게 할 수 있을까요?
`Dto` 를 이용해서 `Line` 을 생성하는 방법이 적절한지 고민해 볼 필요가 있을 것 같습니다.

ssamzag replied:
넵 다른 방법을 찾아보겠습니다.

ssamzag replied:
깔끔하게 LineRequest가 Line까지 생성하도록 만들고 싶었습니다. to 메서드에서 up/down id 끼리 비교해서 예외처리를 할까하다가 그냥 서비스에서 Line 객체를 바로 생성하도록 바꿨습니다.

seokhongkim replied:
이제 사용되지 않는 코드로 보여 삭제하면 좋을 것 같습니다!

ssamzag replied:
넵. 삭제하도록 할게요

seokhongkim commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,63 @@
+        this.stations.add(new StationResponse(downStation.getId(), downStation.getName()));
+    }
+
+    static class StationResponse {
```
내부 클래스로 지정하신 이유가 있을까요? `station` 데이터만 조회하는 경우에도 `LineResponse` 내부에 있는 `StationResponse` 를 사용하지는 않을까요?

ssamzag replied:
요구사항 http 명세서에 날짜 필드가 모두 빠져있어 날짜를 뺀 클래스를 내부에 따로 만들었습니다. 
기존에 있는 StationResponse 클래스 파일의 날짜필드들을 삭제하고 사용하면 될까요?

seokhongkim replied:
제가 1단계 미션에서 `StationResponse` 를 꼼꼼하게 확인하지 못했던 것 같네요. 😅 
1단계 미션 내용을 보니 역 조회 API에서도 날짜 필드는 불필요한 것 같습니다!

seokhongkim commented on src/test/java/nextstep/subway/domain/LineTest.java:
```diff
@@ -0,0 +1,4 @@
+package nextstep.subway.domain;
+
+public class LineTest {
```
테스트 파일만 있고 테스트 코드는 존재하지 않네요. 삭제해도 괜찮지 않을까요?

ssamzag replied:
넵 삭제할게요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,62 @@
+                .collect(Collectors.toList());
+    }
+
+    public LineResponse findById(Long id) {
```
데이터가 없어서 `null` 이 반환되는 케이스가 있을 수 있음을 명시적으로 표현할 수 있는 방법은 어떤 게 있을까요?

ssamzag replied:
위의 예와 마찬가지로 예외를 처리하도록 하겠습니다

seokhongkim reviewed (changes requested):
민호님, ATDD 2단계 미션 잘 구현해주셨습니다. 👍 
문의주신 내용에 대한 답변은 슬랙으로 말씀드려서 생략하겠습니다.

커밋 이력을 보니, 운영 코드를 먼저 커밋해주신 것 같습니다.
혹시 미션을 진행하실 때 운영 코드를 먼저 작성하셨다면, 의식적으로 테스트를 먼저 작성하시는 연습을 해보시면 학습에 도움이 될 것 같습니다.
테스트 먼저 작성 후 커밋 순서만 다르게 진행하신거라면 괜찮습니다. 😀 

코드에 리뷰 피드백을 조금 남겼는데, 확인 부탁드립니다. 🙏 
궁금하신 점이 있다면 깃헙 코멘트 또는 슬랙 DM 으로 편하게 말씀 부탁드립니다.

ssamzag committed:
feat: 파일 추가 및 안 미사용 테스트클래스 삭제

ssamzag committed:
test: 404 에러 테스트

ssamzag committed:
refactor: 리뷰 피드백 반영
- distance 필드 컬럼 설정
- lines -> line 변경
- request parameter 를 단건씩 처리
- 수정/삭제하려는 데이터가 없는 경우 DataNotFoundException 사용
- 노선 생성시 Line 객체 생성 방법 변경

seokhongkim commented on src/main/java/nextstep/subway/application/StationService.java:
```diff
@@ -33,6 +35,14 @@ public List<StationResponse> findAllStations() {
                 .collect(Collectors.toList());
     }
 
+    public Station findStationById(Long id) {
```
레포지토리에서는 `findBy~` 메소드가 값이 없을 경우 `Optional.empty` 객체를 반환하는데, 서비스에서는 동일한 `findBy~` 메소드가 값이 없는 경우 예외를 발생시키네요.
메소드의 이름만 보고도 값이 반드시 존재하거나, 없을 경우 예외가 발생한다는 것을 알 수 있으면 더 좋을 것 같아요!

ssamzag replied:
컨트롤러에서 repository.findby~ 를 바로 호출하려다가 서비스에 만든건데 메서드명부터 좀 애매했던 것 같네요. 컨트롤러에서 예외 처리하게 만들게요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,70 @@
+        lineRepository.delete(line);
+    }
+
+    private Line findById(Long id) {
```
레포지토리에서는 findBy~ 메소드가 값이 없을 경우 Optional.empty 객체를 반환하는데, 서비스에서는 동일한 findBy~ 메소드가 값이 없는 경우 예외를 발생시키네요.
메소드의 이름만 보고도 값이 반드시 존재하거나, 없을 경우 예외가 발생한다는 것을 알 수 있으면 더 좋을 것 같아요!

seokhongkim commented on src/main/java/nextstep/subway/application/LineService.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    @Transactional
+    public LineResponse saveLine(LineRequest lineRequest) {
```
서비스 계층에서 `Line` 을 반환하는 경우도 있고, `LineResponse` 를 반환하는 경우도 있네요.
혹시 현재 `Line`, `LineResponse` 를 반환하는 기준이 다른 이유가 컨트롤러에서 그 값을 사용하는지 여부에 따라 달라지지는 않나요?
서비스 계층에서는 도메인과 DTO 중 어떤 것을 반환하는 게 좋다고 생각하시나요?

ssamzag replied:
서비스에서 DTO까지 만들어 주는 게 깔끔하단 생각이 드는데 공부 좀 해보겠습니다~

seokhongkim commented on src/main/java/nextstep/subway/dto/LineResponse.java:
```diff
@@ -0,0 +1,63 @@
+        this.stations.add(new StationResponse(downStation.getId(), downStation.getName()));
+    }
+//
+//    static class StationResponse {
```
사용하지 않는 코드는 주석 처리하는 것보다 삭제하는 방향으로 작업해주시는 게 좋을 것 같아요.
기존 코드가 필요하면 커밋 로그를 통해 확인할 수 있습니다. 😀

ssamzag replied:
미처 지우질 못 했네요. 😅  삭제 하도록 하겠습니다

seokhongkim commented on src/main/java/nextstep/subway/exception/DataNotFoundException.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code = HttpStatus.NOT_FOUND)
+public class DataNotFoundException extends RuntimeException {
```
필요한 예외를 정의해서 사용하시는 것 좋습니다. 👍

seokhongkim reviewed (approved):
민호님, ATDD 2단계 미션 피드백도 잘 반영해주셨습니다! 👍 

코멘트를 조금 남겼는데, 이 부분은 3단계 미션 진행하면서 고민해보셔도 좋을 것 같습니다.
2단계 PR은 이만 머지하도록 하겠습니다.

3단계 미션에서 뵙겠습니다. 😀
